File: src/main/java/komu/demodel/domain/Module.java
/*
 * Copyright (C) 2006-2010 Juha Komulainen. All rights reserved.
 */
package komu.demodel.domain;

import static java.util.Collections.unmodifiableList;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import komu.demodel.utils.Check;

public abstract class Module {

    private final String name;
    protected final PackageModule parent;
    private final Map<Module,Integer> strengths = new HashMap<Module,Integer>();
    private final Map<Module,List<Dependency>> dependencies = new HashMap<Module,List<Dependency>>();
    
    private final Map<Module,Integer> cachedDependencyStrengths = new HashMap<Module,Integer>();
    private List<Module> cachedSelfAndAncestors = null;
    
    public Module(String name, PackageModule parent) {
        Check.notNull(name, "name");
        
        this.name = name;
        this.parent = parent;
        if (parent != null)
            parent.addChild(this);
    }
    
    public abstract ModuleType getType();
    public abstract List<Module> getChildren();
    public abstract void sortChildren();
    
    abstract PackageModule getRoot();
    
    public void flushCaches() {
        cachedDependencyStrengths.clear();
        cachedSelfAndAncestors = null;
        
        for (Module child : getChildren())
            child.flushCaches();
    }
    
    public int getDepth() {
        return (parent != null) ? parent.getDepth() + 1 : 0;
    }
    
    public Module getParent() {
        return parent;
    }
    
    public String getName() {
        return name;
    }
    
    public String getLocalName() {
        if (parent == null) return name;
        
        return name.startsWith(parent.getName() + ".")
                ? name.substring(parent.getName().length() + 1)
                : name;
    }
    
    public abstract boolean isProgramModule();
    
    public abstract void filterNonProgramReferences();
    
    public abstract boolean isLeaf();
    public abstract int countLeafsUnder();
    
    public int getDependencyStrength(Module module) {
        Integer cachedStrength = cachedDependencyStrengths.get(module);
        if (cachedStrength != null) {
            return cachedStrength.intValue();
        } else {
            int strength = getUncachedDependencyStrength(module);
            cachedDependencyStrengths.put(module, strength);
            return strength;
        }
    }

    private int getUncachedDependencyStrength(Module module) {
        List<Module> myAncestors = getSelfAndAncestors();
        if (myAncestors.contains(module)) return 0;
            
        List<Module> targetAncestors = module.getSelfAndAncestors();
        if (targetAncestors.contains(this)) return 0;
            
        int strength = 0;
        for (Module ancestor : myAncestors)
            for (Module target : targetAncestors)
                strength += ancestor.getDirectDependencyStrength(target);

        return strength;
    }
    
    public List<Module> getSelfAndAncestors() {
        if (cachedSelfAndAncestors == null) {
            List<Module> result = new ArrayList<Module>();
            result.add(this);
            addAncestors(result);
            cachedSelfAndAncestors = unmodifiableList(result);
        }
        return cachedSelfAndAncestors;
    }
    
    private void addAncestors(List<Module> result) {
        for (Module child : getChildren()) {
            result.add(child);
            child.addAncestors(result);
        }
    }

    int getDirectDependencyStrength(Module module) {
        Integer strength = strengths.get(module);
        return (strength != null) ? strength.intValue() : 0;
    }

    public void setDependencyStrength(Module module, int strength) {
        Check.notNull(module, "module");
        Check.notNegative(strength, "strength");
        
        strengths.put(module, strength);
    }

    public void addDependency(Module module, DependencyType type) {
        Check.notNull(module, "module");
        Check.notNull(type, "type");
        
        Integer strength = strengths.get(module);
        strengths.put(module, (strength != null) ? (strength + 1) : 1);
        
        // TODO: We get dependencies to void, java.lang.Class, java.lang.String as well, maybe we should get rid of them
        List<Dependency> list = dependencies.get(module);
        if (list == null) {
            list = new ArrayList<Dependency>();
            dependencies.put(module, list);
        }
        list.add(new Dependency(module, type));
    }
    
    public List<Dependency> getDependencies(Module to) {
        return dependencies.get(to);
    }
    
    @Override
    public String toString() {
        return name;
    }
    
    public void move(MoveDirection direction) {
        if (parent != null)
            parent.move(this, direction);
    }
    
    int outgoingDependencies(Collection<Module> modules) {
        int sum = 0;
        for (Module m : modules)
            sum += getDependencyStrength(m);
        return sum;
    }
}

File: src/main/scala/komu/demodel/domain/Module.scala
/*
 * Copyright (C) 2006-2010 Juha Komulainen. All rights reserved.
 */
package komu.demodel.domain

import collection.mutable.{ ArrayBuffer, HashMap }

abstract class Module(val name: String, val parent: Option[PackageModule]) {

  for (p <- parent)
    p.addChild(this)

  private val strengths = new HashMap[Module,Int]
  private val dependencies = new HashMap[Module,List[Dependency]]
  private val cachedDependencyStrengths = new HashMap[Module,Int]
  private var cachedSelfAndAncestors: List[Module] = null

  def children: List[Module]
  def root: PackageModule
  def moduleType: ModuleType
  def sortChildren()
  def isProgramModule: Boolean
  def filterNonProgramReferences()
  def isLeaf: Boolean
  def countLeafsUnder: Int

  def flushCaches() {
    cachedDependencyStrengths.clear();
    cachedSelfAndAncestors = null;
        
    for (child <- children) child.flushCaches()
  }
    
  def depth: Int =
    parent match {
      case Some(p) => p.depth + 1
      case None    => 0
    }

  def localName =
    parent match {
      case None    => name
      case Some(p) => if (name.startsWith(p.name + "."))
                        name.substring(p.name.length+1)
                      else
                        name
    }

  def getDependencyStrength(module: Module): Int =
    cachedDependencyStrengths.getOrElseUpdate(module, getUncachedDependencyStrength(module))

  private def getUncachedDependencyStrength(module: Module): Int = {
    val myAncestors = selfAndAncestors
    if (myAncestors.contains(module)) return 0
            
    val targetAncestors = module.selfAndAncestors
    if (targetAncestors.contains(this)) return 0
            
    var strength = 0
    for (ancestor <- myAncestors)
      for (target <- targetAncestors)
        strength += ancestor.getDirectDependencyStrength(target)
    strength
  }
    
  def selfAndAncestors = {
    if (cachedSelfAndAncestors == null) {
        val result = new ArrayBuffer[Module]
        result += this
        addAncestors(result)
        cachedSelfAndAncestors = result.toList
    }
    cachedSelfAndAncestors;
  }
    
  private def addAncestors(result: ArrayBuffer[Module]) {
    for (child <- children) {
      result += child
      child.addAncestors(result)
    }
  }

  def getDirectDependencyStrength(module: Module) =
    strengths.getOrElse(module, 0)

  def setDependencyStrength(module: Module, strength: Int) {
    assert(module != null)
    assert(strength >= 0)
        
    strengths.put(module, strength)
  }

  def addDependency(module: Module, kind: DependencyType) {
    assert(module != null)
    assert(kind != null)
        
    strengths(module) = getDirectDependencyStrength(module) + 1
    dependencies(module) = new Dependency(module, kind) :: getDependencies(module)
  }
  
  def getDependencies(to: Module) =
    dependencies.getOrElse(to, Nil)

  override def toString = name

  def move(direction: MoveDirection) {
    for (p <- parent)
      p.move(this, direction)
  }

  def outgoingDependencies(modules: Iterable[Module]) =
    modules.map(getDependencyStrength(_)).sum
}
